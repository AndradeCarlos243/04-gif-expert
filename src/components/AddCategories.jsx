import { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ handleAddCategory }) => {
    const [ txtCategory, setTxtCategory ] = useState('');

    const handleChangeCategory = ({target: {value}}) => {
        setTxtCategory(value);
    }

    const handleFormAdd = (e) => {
        e.preventDefault();
        const category = txtCategory.trim();
        if( category.length <= 2 ) return false;
        handleAddCategory(category);
        setTxtCategory('');
    }

    return (
        <>
            <form onSubmit={ handleFormAdd } aria-label='form'>
                <input type="text" name="txt-category" id="txt-category" value={ txtCategory } placeholder='Buscar Gifs' onChange={ handleChangeCategory }/>
            </form>
        </>
    );
}

export default AddCategory;

AddCategory.propTypes = {
    handleAddCategory: PropTypes.func.isRequired,
}