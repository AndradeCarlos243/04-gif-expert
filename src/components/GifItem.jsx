import { PropTypes } from 'prop-types';

export const GifItem = ({ url, title, description }) => {
  return (
    <div className='card'>
        <div className='img-container'>
            <img src={url} alt={description} width={300} height={300} style={{objectFit: 'cover'}}/>
        </div>
        {/* <div className='card-tags'>
            {
                tags.map( (tag, i) => 
                {
                    if(i >= 3) return false;
                    return <span key={tag.replace(/ /g, "_")} className='tag'>{tag.replace(/ /g, "_")}</span>
                })
            }
        </div> */}
        <p>{title}</p>
    </div>
  )
}

GifItem.propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired
}