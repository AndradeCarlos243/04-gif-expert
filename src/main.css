
* {
    font-family: Helvetica, Arial, sans-serif;
    background-color: rgb(234, 234, 234);
}
  
body {
    padding: 60px;
}

input {
    background-color: white;
    border-radius: 5px;
    border: 1px solid rgb(97, 32, 158);
    color: black;
    font-size: 1.2rem;
    outline: none;
    padding: 10px 15px;
    width: 100%;
}

h2 {
    font-size: 1.5rem;
}

h3 {
    font-size: 3rem;
    margin-bottom: 5px;
}

.card-grid {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    margin: 2rem;
}

.card {
    align-content: center;
    align-items: center;
    background-color: white;
    border-radius: 10px;
    box-shadow: 0px 3px 5px rgba(0, 0, 0, 0.03);
    display: flex;
    flex-direction: column; 
    justify-content: center;
    margin-bottom: 20px;
    margin-right: 20px;
    overflow: hidden;
    width: 10rem;
}

.card p {
    background-color: white;
    flex: 1;
    font-size: .9rem;
    margin-top: 5px;
    padding: 5px 20px 0px 20px;
    text-align: start;
}

.card .img-container {
    background-color: white;
    width: 10rem;
    height: 10rem;
}

.card img {
    object-fit: cover;
    width: 10rem;
    height: 10rem;
}

.card .tag {
    background-color: rgba(107, 113, 185, 0.74);
    color: white;
    font-size: .8rem;
    padding: .3rem 1rem;
    margin: 0.1rem;
    border-radius: 1rem;
    overflow: hidden;
    position: relative;
    display: inline-block;
    text-align: center;
    text-decoration: none;
    text-overflow: ellipsis;
    white-space: nowrap;
    max-width: 7rem;
}

.card .tag:hover {
    background-color: rgba(107, 113, 185, 1);
    cursor: pointer;
}

.card .card-tags {
    justify-content: center;
    background-color: white;
    display: flex;
    flex-wrap: wrap;
    padding: 1rem;
}

.logo-tenor
{
    width: 5rem;
    margin: .5rem 0;
}