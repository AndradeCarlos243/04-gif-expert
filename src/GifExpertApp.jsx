import { AddCategory, GifGrid } from './components';
import { useState } from 'react';
import { tenorLogo } from './assets/svg/via_tenor_logo_grey.svg';

export const GifExpertApp = () => {
    const [ categories, setCategories ] = useState(['K-On! anime', 'Frieren: Beyond Journey\'s End', 'Kobayashi-san Chi no Maid Dragon']);

    const handleAddCategory = ( category ) => {
        if( categories.includes(category) ) return false;
        setCategories([category, ...categories]);
    }

    return (
        <>
            <h1>{"Nasemi Gif's App".toUpperCase()} <img className="logo-tenor" src={tenorLogo} alt="marca de agua de tenor" width={100} /></h1>
            <AddCategory handleAddCategory = { handleAddCategory } />
            {
                categories.map( category => (
                    <GifGrid key={ category.replace(/ /g, "_") } category={ category } query={category.replace(/ /g, "-").toLowerCase()} />
                ))    
            }
        </>
    );
}

export default GifExpertApp;