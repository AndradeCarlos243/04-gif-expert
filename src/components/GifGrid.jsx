import PropTypes from 'prop-types';
import { GifItem } from '../components';
import { useFetchGifs } from '../hooks';

export const GifGrid = ({ query, category }) => {
    const { images, isLoading } = useFetchGifs(query);

    return (
        <>
            <h3>
                { category }
            </h3>
            { isLoading && <p>Cargando...</p> }
            <div className="card-grid">
                {
                    images.map( (image) => (
                        <GifItem 
                        key={ image.id }
                        {... image }
                        />
                    ))
                }
            </div>
        </>
    )
}

export default GifGrid;

GifGrid.propTypes = {
    query: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired
}