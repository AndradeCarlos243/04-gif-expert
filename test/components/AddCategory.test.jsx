import { fireEvent, render, screen } from '@testing-library/react';
import { AddCategory } from '../../src/components';

describe('Pruebas en AddCategory', () => {
    const category = 'eromanga sensei';
    const handleAddCategory = jest.fn();

    test('Debe hacer match con el snapshot', () => {
        const { container } = render(<AddCategory handleAddCategory={() => {}} />);
        expect(container).toMatchSnapshot();
    })

    test('Debe cambiar la caja de texto', () => {
        render(<AddCategory handleAddCategory={() => {}} />);
        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { value: category } });
        expect(input.value).toBe(category);
    });

    test('No debe postear la informaciÃ³n con submit', () => {
        render(<AddCategory handleAddCategory={ handleAddCategory } />);
        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');
        fireEvent.input(input, { target: { value: '' } });
        fireEvent.submit(form);
        expect(handleAddCategory).not.toHaveBeenCalled();
    });

    test('Debe llamar el handleAddCategory y limpiar la caja de texto', () => {
        render(<AddCategory handleAddCategory={ handleAddCategory } />);
        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');
        fireEvent.input(input, { target: { value: category } });
        fireEvent.submit(form);
        expect(handleAddCategory).toHaveBeenCalledTimes(1);
        expect(handleAddCategory).toHaveBeenCalledWith( category );
        expect(input.value).toBe('');
    });
});