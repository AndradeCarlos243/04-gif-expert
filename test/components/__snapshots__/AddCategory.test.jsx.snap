// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddCategory Debe hacer match con el snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      id="txt-category"
      name="txt-category"
      placeholder="Buscar Gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
